bcda/auth/oktabackend.go

```diff
@@ -0,0 +1,321 @@
+		return err
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("Authorization", authString)
+
+	var client = &http.Client{Timeout: time.Second * 10,}
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	var result []Policy
+	err = json.NewDecoder(resp.Body).Decode(&result)
+	if err != nil {
+		return err
+	}
+
+	if len(result) > 1 {
+		return fmt.Errorf("more than one policy entry for server; can't continue safely")
+	}
+
+	// add the new client to the inclusion list
+	incl := result[0].Conditions.Clients.Include
+	incl = append(incl, clientID)
+	result[0].Conditions.Clients.Include = incl
+
+	// put the list back to the server
+	body, err := json.Marshal(result[0])
+	if err != nil {
+		return err
+	}
+
+	req, err = http.NewRequest("PUT", fmt.Sprintf("%s/%s", policyUrl, result[0].ID), bytes.NewBuffer(body))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("Authorization", authString)
+
+	client = &http.Client{Timeout: time.Second * 10,}
+	resp, err = client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != 200 {
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			log.Printf("failed to update policy with %s because %v", clientID, err)
+			return err
+		}
+		return fmt.Errorf("unexpected result: %s", body)
+	}
+
+	return nil
+}
+
+func refreshKeys() {
+	for range time.Tick(time.Hour * 1) {
+		log.Println("refreshing okta public keys")
+		getPublicKeys()
+	}
+}
+
+// Gets the current set of public server signing keys. This code treats all failures as fatal,
+// since failures are likely to result in an unrecoverable state for the application.
+func getPublicKeys() map[string]rsa.PublicKey {
+	var (
+		body []byte
+		err  error
+		keys map[string]rsa.PublicKey
+	)
+
+	body, err = get(os.Getenv("OKTA_OAUTH_SERVER_METADATA"))
+	if err != nil {
+		log.Errorf("can't fetch okta oauth server metadata because %s", err)
+	}
+	md := make(map[string]interface{})
+	if err = json.Unmarshal(body, &md); err != nil {
+		log.Errorf("can't unmarshal %s because %s", string(body), err)
+	}
+
+	jwkUrl := md["jwks_uri"].(string)
+	body, err = get(jwkUrl)
+	if err != nil {
+		log.Errorf("can't get keys from %s because %s", jwkUrl, err)
+	}
+
+	keys, err = parseKeys(body)
+	if err != nil {
+		log.Errorf("couldn't parse %s because %s", string(body), err)
+	}
+	if len(keys) == 0 {
+		log.Fatal("must have at least 1 public key; have none")
+	}
+
+	log.Printf("%d okta public oauth server public keys cached", len(keys))
+	return keys
+}
+
+type RsaJWK struct {
+	KeyType   string `json:"kty"`
+	Algorithm string `json:"alg"`
+	ID        string `json:"kid"`
+	Use       string `json:"use"`
+	E         string `json:"e"`
+	N         string `json:"n"`
+}
+
+type KeyList struct {
+	Keys []*RsaJWK `json:"keys"`
+}
+
+func get(urlString string) ([]byte, error) {
```

Comment:
also think about moving this function and the places it is called to the `oktaclient.go`.